languageServer/src/languageHelpers/baseLanguageHelper.ts

```diff
@@ -0,0 +1,105 @@
+import { has } from 'lodash'
+import { TextDocument } from 'vscode-languageserver-textdocument'
+import {
+  DocumentSymbol,
+  Position,
+  Location,
+  SymbolKind
+} from 'vscode-languageserver/node'
+
+export interface LanguageHelper {
+  findSymbolAtPosition(position: Position): DocumentSymbol | undefined
+  findLocationsFor(symbol: DocumentSymbol): Location[]
+}
+
+export abstract class BaseLanguageHelper<RootNode> implements LanguageHelper {
+  protected textDocument: TextDocument
+  protected rootNode?: RootNode
+
+  constructor(textDocument: TextDocument) {
+    this.textDocument = textDocument
+    this.rootNode = this.parse(this.getText())
+  }
+
+  public findSymbolAtPosition(position: Position): DocumentSymbol | undefined {
+    const cursorOffset: number = this.offsetAt(position)
+    const symbolsAroundOffset = this.findEnclosingSymbols(cursorOffset)
+
+    const symbolStack = symbolsAroundOffset.sort((a, b) => {
+      const offA = this.offsetAt(a.range.end) - this.offsetAt(a.range.start)
+      const offB = this.offsetAt(b.range.end) - this.offsetAt(b.range.start)
+
+      return offB - offA // We want the tighter fits for last, so we can just pop them
+    })
+
+    return [...symbolStack].pop()
+  }
+
+  public findLocationsFor(symbol: DocumentSymbol): Location[] {
+    if (symbol.kind === SymbolKind.Property) {
+      return this.findLocationsForPropertySymbol(symbol)
+    }
+
+    return this.findLocationsForNormalSymbol(symbol)
+  }
+
+  protected getText() {
+    return this.textDocument.getText()
+  }
+
+  protected offsetAt(position: Position) {
+    return this.textDocument.offsetAt(position)
+  }
+
+  protected positionAt(offset: number) {
+    return this.textDocument.positionAt(offset)
+  }
+
+  private findLocationsForPropertySymbol(symbol: DocumentSymbol) {
+    const propertyPath = symbol.detail
+    const itIsHere = propertyPath && this.hasProperty(propertyPath)
```

Comment:
Nitpick, but is it possible to rename `isHere` as it follows our boolean naming conventions more.

Reply:
Yes, thanks!